### ISSUE TITLE
Allow for async filters

### ISSUE THREAD

**Issue by king-of-poppk** [rstudio/plumber#907]
With #204 implemented we now have a pretty convenient way to parallelize IO resource consumption. However, it seems plumber filters still do not have this feature. Would it be possible to make it so that filters can return promises for the `plumber::forward()` call?

**Comment by meztez** [comment]
@king-of-poppk how would it be used?

Why would it not work currently? Aren't filters are just regular steps according to runSteps source?

**Comment by king-of-poppk** [comment]
@meztez I may be wrong. I will double-check.

**Comment by meztez** [comment]
@king-of-poppk I did some further testing and I do not think plumber::forward could be used within a future::future. What would be the use case?

**Comment by schloerke** [comment]
Correct. 

You can have a future return and then call forward() within the follow up promise. 

---------------

So if the filter could return a promise and within that promise (which executes in the main R worker) it calls forward(), I think it'll work.

**Comment by meztez** [comment]
```r
library(plumber)
library(future)
plan("multisession")

#* @filter future
function(req) {
  print(req$PATH_INFO)
  if (req$PATH_INFO %in% c("/__docs__/", "/openapi.json")) {
    plumber::forward()
  }
  res <- NULL
  future::future({
    if (req$PATH_INFO=="/echo1") {
      Sys.sleep(10)
    }
    plumber::forward()
  })
}

#* @serializer text
#* @get /echo1
function(msg) {
  msg
}

#* @serializer text
#* @get /echo2
function(msg) {
  msg
}

```

I'm working from this, only getting a json serialized true. I tweak some things from what I understand from your comment.

**Comment by schloerke** [comment]
I'm on mobile, so formatting will not be good / tested. 

Updating your future code to something like this...

```r
  future::future({
    if (req$PATH_INFO=="/echo1") {
      Sys.sleep(10)
    }
  }) %...>% {
    value <- .

    plumber::forward()
    value
  }
```

(I hope I'm not making stuff up here 😅)

------------

I believe you could even call forward before the promise is returned. 

```r
  plumber::forward()
  future::future({
    if (req$PATH_INFO=="/echo1") {
      Sys.sleep(10)
    }
    # return data here
  })
```

**Comment by schloerke** [comment]
Important part is that `forward()` is called within the main R worker.

**Comment by meztez** [comment]
@schloerke Oh, I get it. I thought it needed to be called last but it just sets a flag. Ok, this will work now.

**Comment by schloerke** [comment]
I wish `forward()` was required to be returned last. 

Migrating towards that goal is really hard without spooky deprecation message or a separate solution entirely (ex: endpoint based filters).

**Comment by meztez** [comment]
Yep it works @king-of-poppk , let me know if this is something you can use.

```r
library(plumber)
library(future)
plan("multisession")

#* @filter future
function(req) {
  plumber::forward()
  future::future({
    if (req$PATH_INFO=="/echo1") {
      Sys.sleep(10)
    }
  })
}

#* @serializer text
#* @get /echo1
function(msg) { msg }

#* @serializer text
#* @get /echo2
function(msg) { msg }

```

**Comment by meztez** [comment]
> Important part is that `forward()` is called within the main R worker.

Yeah, because it sets the flags in the .global env?

**Comment by schloerke** [comment]
> > Important part is that `forward()` is called within the main R worker.
> 
> 
> Yeah, because it sets the flags in the .global env?

Yup!

**Comment by king-of-poppk** [comment]
> Can you show me a small copy of this code? The `async`/`await` logic is built on a state machine which execute in the main R process.

Ah yes of course my mistake, it should have worked then.

**Comment by thomasp85** [comment]
Closing this as the discussion has been stale for +1 year
